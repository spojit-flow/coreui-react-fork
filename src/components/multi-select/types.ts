export type Option = {
  custom?: boolean
  disabled?: boolean
  selected?: boolean
  text: string // TODO: change to label in v5
  value: number | string
  [key: string]: number | string | any
}

export type OptionsGroup = {
  label: string
  options?: Option[]
  [key: string]: number | string | any
}

export type SelectedOption = {
  disabled?: boolean
  text: string
  value: number | string
  [key: string]: number | string | any
}
