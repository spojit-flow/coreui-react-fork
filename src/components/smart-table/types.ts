import { ReactNode } from 'react'

import { CTableDataCellProps } from '../table/CTableDataCell'
import { CTableHeaderCellProps } from '../table/CTableHeaderCell'
import { CTableRowProps } from '../table/CTableRow'

export type Column = {
  children?: Column[]
  filter?: boolean | ((values: any[], onChange: (value: any) => void) => ReactNode)
  group?: string
  key: string
  label?: string
  sorter?: boolean
  _style?: any
  _props?: CTableHeaderCellProps
}

export type ColumnFilter = {
  lazy?: boolean
  external?: boolean
}

export type ColumnFilterValue = {
  [key: string]: any
}

export type FooterItem = {
  label?: string
  _props?: CTableDataCellProps
}

export type Group = {
  children?: Group[] | Column[]
  colspan?: number
  deep?: number
  group?: string
  key: string
  label?: string
  _style?: any
  _props?: CTableHeaderCellProps
}

export type Item = {
  [key: string]: number | string | any
  _cellProps?: any
  _props?: CTableRowProps
  _selected?: boolean
}

export type ItemsPerPageSelect = {
  external?: boolean
  label?: string
  values?: Array<number>
}

export type Pagination = {
  external?: boolean
}

export type Sorter = {
  resetable?: boolean
  external?: boolean
}

export type ScopedColumns = {
  [key: string]: any
  details?: (a: Item, b: number) => ReactNode
}

export type SorterValue = {
  column?: string
  state?: number | string
}

export type TableFilter = {
  lazy?: boolean
  external?: boolean
}
