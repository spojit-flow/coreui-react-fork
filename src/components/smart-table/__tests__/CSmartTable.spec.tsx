import * as React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { CBadge, CSmartTable } from '../../../index'
import { Item } from '../types'

test('loads and displays CSpinner component', async () => {
  const columns = [
    {
      key: 'name',
      _style: { width: '20%', color: 'blue' },
      _props: { color: 'primary', className: 'fw-semibold' },
    },
    'registered',
    { key: 'role', filter: false, sorter: false, _style: { width: '20%' } },
    { key: 'status', _style: { width: '20%' } },
    {
      key: 'show_details',
      label: '',
      _style: { width: '1%' },
      filter: false,
      sorter: false,
      _props: { color: 'primary', className: 'fw-semibold' },
    },
  ]
  const usersData = [
    { id: 0, name: 'John Doe', registered: '2022/01/01', role: 'Guest', status: 'Pending' },
    {
      id: 1,
      name: 'Samppa Nori',
      registered: '2022/01/01',
      role: 'Member',
      status: 'Active',
      _props: { color: 'primary', align: 'middle' },
    },
    {
      id: 2,
      name: 'Estavan Lykos',
      registered: '2022/02/07',
      role: 'Staff',
      status: 'Banned',
      _cellProps: { all: { className: 'fw-semibold' }, name: { color: 'info' } },
    },
    { id: 3, name: 'Chetan Mohamed', registered: '2022/02/07', role: 'Admin', status: 'Inactive' },
    {
      id: 4,
      name: 'Derick Maximinus',
      registered: '2022/03/19',
      role: 'Member',
      status: 'Pending',
    },
    { id: 5, name: 'Friderik Dávid', registered: '2022/01/21', role: 'Staff', status: 'Active' },
    { id: 6, name: 'Yiorgos Avraamu', registered: '2022/01/01', role: 'Member', status: 'Active' },
    {
      id: 7,
      name: 'Avram Tarasios',
      registered: '2022/02/07',
      role: 'Staff',
      status: 'Banned',
      _props: { color: 'warning', align: 'middle' },
    },
    { id: 8, name: 'Quintin Ed', registered: '2022/02/07', role: 'Admin', status: 'Inactive' },
    { id: 9, name: 'Enéas Kwadwo', registered: '2022/03/19', role: 'Member', status: 'Pending' },
    { id: 10, name: 'Agapetus Tadeáš', registered: '2022/01/21', role: 'Staff', status: 'Active' },
    { id: 11, name: 'Carwyn Fachtna', registered: '2022/01/01', role: 'Member', status: 'Active' },
    { id: 12, name: 'Nehemiah Tatius', registered: '2022/02/07', role: 'Staff', status: 'Banned' },
    { id: 13, name: 'Ebbe Gemariah', registered: '2022/02/07', role: 'Admin', status: 'Inactive' },
    {
      id: 14,
      name: 'Eustorgios Amulius',
      registered: '2022/03/19',
      role: 'Member',
      status: 'Pending',
    },
    { id: 15, name: 'Leopold Gáspár', registered: '2022/01/21', role: 'Staff', status: 'Active' },
    { id: 16, name: 'Pompeius René', registered: '2022/01/01', role: 'Member', status: 'Active' },
    { id: 17, name: 'Paĉjo Jadon', registered: '2022/02/07', role: 'Staff', status: 'Banned' },
    {
      id: 18,
      name: 'Micheal Mercurius',
      registered: '2022/02/07',
      role: 'Admin',
      status: 'Inactive',
    },
    {
      id: 19,
      name: 'Ganesha Dubhghall',
      registered: '2022/03/19',
      role: 'Member',
      status: 'Pending',
    },
    { id: 20, name: 'Hiroto Šimun', registered: '2022/01/21', role: 'Staff', status: 'Active' },
    { id: 21, name: 'Vishnu Serghei', registered: '2022/01/01', role: 'Member', status: 'Active' },
    { id: 22, name: 'Zbyněk Phoibos', registered: '2022/02/07', role: 'Staff', status: 'Banned' },
    {
      id: 23,
      name: 'Aulus Agmundr',
      registered: '2022/01/01',
      role: 'Member',
      status: 'Pending',
      _selected: true,
    },
    {
      id: 42,
      name: 'Ford Prefect',
      registered: '2001/05/25',
      role: 'Alien',
      status: "Don't panic!",
    },
  ]
  const getBadge = (status: string) => {
    switch (status) {
      case 'Active':
        return 'success'
      case 'Inactive':
        return 'secondary'
      case 'Pending':
        return 'warning'
      case 'Banned':
        return 'danger'
      default:
        return 'primary'
    }
  }
  const { container } = render(
    <CSmartTable
      activePage={3}
      cleaner
      clickableRows
      columns={columns}
      columnFilter
      columnSorter
      footer
      items={usersData}
      itemsPerPageSelect
      itemsPerPage={5}
      pagination
      scopedColumns={{
        status: (item: Item) => (
          <td>
            <CBadge color={getBadge(item.status)}>{item.status}</CBadge>
          </td>
        ),
      }}
      selectable
      sorterValue={{ column: 'status', state: 'asc' }}
      tableFilter
      tableHeadProps={{
        color: 'danger',
      }}
      tableProps={{
        className: 'add-this-class',
        responsive: true,
        striped: true,
        hover: true,
      }}
    />,
  )
  expect(container).toMatchSnapshot()
})
