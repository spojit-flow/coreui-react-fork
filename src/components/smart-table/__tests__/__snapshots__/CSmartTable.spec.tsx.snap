// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and displays CSpinner component 1`] = `
<div>
  <div>
    <div
      class="row my-2 mx-0"
    >
      <div
        class="col-auto p-0"
      >
        <div
          class="row mb-2"
        >
          <label
            class="form-label col-sm-auto col-form-label"
          >
            Filter:
          </label>
          <div
            class="col-sm-auto"
          >
            <input
              class="form-control"
              placeholder="type string..."
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="col-auto p-0"
      >
        <button
          class="btn btn-transparent"
          type="button"
        >
          <svg
            class="icon icon-custom-size"
            role="img"
            viewBox="0 0 512 512"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              class="ci-primary"
              fill="var(--ci-primary-color, currentColor)"
              points="40 16 40 53.828 109.024 136 150.815 136 76.896 48 459.51 48 304 242.388 304 401.373 241.373 464 240 464 240 368 208 368 208 496 254.627 496 336 414.627 336 253.612 496 53.612 496 16 40 16"
            />
            <polygon
              class="ci-primary"
              fill="var(--ci-primary-color, currentColor)"
              points="166.403 248.225 226.864 187.763 204.237 165.135 143.775 225.597 83.313 165.135 60.687 187.763 121.148 248.225 60.687 308.687 83.313 331.314 143.775 270.852 204.237 331.314 226.864 308.687 166.403 248.225"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="position-relative"
  >
    <div
      class="table-responsive"
    >
      <table
        class="table table-hover table-striped add-this-class"
      >
        <thead
          class="table-danger"
        >
          <tr>
            <th>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </th>
            <th
              class="table-primary fw-semibold"
              style="cursor: pointer; width: 20%; color: blue;"
            >
              <div
                class="d-inline"
              >
                Name
              </div>
              <span
                class="opacity-25 float-end me-1"
              >
                <svg
                  class="icon icon-custom-size"
                  role="img"
                  viewBox="0 0 512 512"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    class="ci-primary"
                    fill="var(--ci-primary-color, currentColor)"
                    points="384 433.373 384 160 352 160 352 434.51 282.177 364.687 259.55 387.313 367.432 495.196 475.313 387.313 452.687 364.687 384 433.373"
                  />
                  <polygon
                    class="ci-primary"
                    fill="var(--ci-primary-color, currentColor)"
                    points="159.432 17.372 51.55 125.255 74.177 147.882 144 78.059 144 352 176 352 176 79.195 244.687 147.882 267.313 125.255 159.432 17.372"
                  />
                </svg>
              </span>
            </th>
            <th
              style="cursor: pointer;"
            >
              <div
                class="d-inline"
              >
                Registered
              </div>
              <span
                class="opacity-25 float-end me-1"
              >
                <svg
                  class="icon icon-custom-size"
                  role="img"
                  viewBox="0 0 512 512"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    class="ci-primary"
                    fill="var(--ci-primary-color, currentColor)"
                    points="384 433.373 384 160 352 160 352 434.51 282.177 364.687 259.55 387.313 367.432 495.196 475.313 387.313 452.687 364.687 384 433.373"
                  />
                  <polygon
                    class="ci-primary"
                    fill="var(--ci-primary-color, currentColor)"
                    points="159.432 17.372 51.55 125.255 74.177 147.882 144 78.059 144 352 176 352 176 79.195 244.687 147.882 267.313 125.255 159.432 17.372"
                  />
                </svg>
              </span>
            </th>
            <th
              style="width: 20%;"
            >
              <div
                class="d-inline"
              >
                Role
              </div>
            </th>
            <th
              style="cursor: pointer; width: 20%;"
            >
              <div
                class="d-inline"
              >
                Status
              </div>
              <span
                class="float-end me-1"
              >
                <svg
                  class="icon icon-custom-size"
                  role="img"
                  viewBox="0 0 512 512"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    class="ci-primary"
                    fill="var(--ci-primary-color, currentColor)"
                    points="390.624 150.625 256 16 121.376 150.625 144.004 173.252 240.001 77.254 240.001 495.236 272.001 495.236 272.001 77.257 367.996 173.252 390.624 150.625"
                  />
                </svg>
              </span>
            </th>
            <th
              class="table-primary fw-semibold"
              style="width: 1%;"
            >
              <div
                class="d-inline"
              />
            </th>
          </tr>
          <tr>
            <th />
            <th
              class="table-primary fw-semibold"
            >
              <input
                aria-label="column name: 'Name' filter input"
                class="form-control form-control-sm"
                type="text"
                value=""
              />
            </th>
            <th>
              <input
                aria-label="column name: 'Registered' filter input"
                class="form-control form-control-sm"
                type="text"
                value=""
              />
            </th>
            <th />
            <th>
              <input
                aria-label="column name: 'Status' filter input"
                class="form-control form-control-sm"
                type="text"
                value=""
              />
            </th>
            <th
              class="table-primary fw-semibold"
            />
          </tr>
        </thead>
        <tbody
          style="cursor: pointer;"
        >
          <tr
            class="align-middle table-warning"
            tabindex="0"
          >
            <td>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </td>
            <td>
              Avram Tarasios
            </td>
            <td>
              2022/02/07
            </td>
            <td>
              Staff
            </td>
            <td>
              <span
                class="badge bg-danger"
              >
                Banned
              </span>
            </td>
            <td>
              undefined
            </td>
          </tr>
          <tr
            tabindex="0"
          >
            <td>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </td>
            <td>
              Nehemiah Tatius
            </td>
            <td>
              2022/02/07
            </td>
            <td>
              Staff
            </td>
            <td>
              <span
                class="badge bg-danger"
              >
                Banned
              </span>
            </td>
            <td>
              undefined
            </td>
          </tr>
          <tr
            tabindex="0"
          >
            <td>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </td>
            <td>
              Paĉjo Jadon
            </td>
            <td>
              2022/02/07
            </td>
            <td>
              Staff
            </td>
            <td>
              <span
                class="badge bg-danger"
              >
                Banned
              </span>
            </td>
            <td>
              undefined
            </td>
          </tr>
          <tr
            tabindex="0"
          >
            <td>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </td>
            <td>
              Zbyněk Phoibos
            </td>
            <td>
              2022/02/07
            </td>
            <td>
              Staff
            </td>
            <td>
              <span
                class="badge bg-danger"
              >
                Banned
              </span>
            </td>
            <td>
              undefined
            </td>
          </tr>
          <tr
            tabindex="0"
          >
            <td>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </td>
            <td>
              Ford Prefect
            </td>
            <td>
              2001/05/25
            </td>
            <td>
              Alien
            </td>
            <td>
              <span
                class="badge bg-primary"
              >
                Don't panic!
              </span>
            </td>
            <td>
              undefined
            </td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <th>
              <input
                class="form-check-input"
                type="checkbox"
              />
            </th>
            <th
              class="table-primary fw-semibold"
              style="width: 20%; color: blue;"
            >
              <div
                class="d-inline"
              >
                Name
              </div>
            </th>
            <th>
              <div
                class="d-inline"
              >
                Registered
              </div>
            </th>
            <th
              style="width: 20%;"
            >
              <div
                class="d-inline"
              >
                Role
              </div>
            </th>
            <th
              style="width: 20%;"
            >
              <div
                class="d-inline"
              >
                Status
              </div>
            </th>
            <th
              class="table-primary fw-semibold"
              style="width: 1%;"
            >
              <div
                class="d-inline"
              />
            </th>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <nav
        aria-label="pagination"
      >
        <ul
          class="pagination justify-content-start"
        >
          <li
            class="page-item"
          >
            <a
              aria-disabled="false"
              aria-label="Go to first page"
              class="page-link"
            >
              «
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-disabled="false"
              aria-label="Go to previous page"
              class="page-link"
            >
              ‹
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-label="Go to page 1"
              class="page-link"
            >
              1
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-label="Go to page 2"
              class="page-link"
            >
              2
            </a>
          </li>
          <li
            aria-current="page"
            class="page-item active"
          >
            <span
              class="page-link"
            >
              3
            </span>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-label="Go to page 4"
              class="page-link"
            >
              4
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-label="Go to page 5"
              class="page-link"
            >
              5
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-disabled="false"
              aria-label="Go to next page"
              class="page-link"
            >
              ›
            </a>
          </li>
          <li
            class="page-item"
          >
            <a
              aria-disabled="false"
              aria-label="Go to last page"
              class="page-link"
            >
              »
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      class="col-auto ms-auto"
    >
      <div
        class="row"
      >
        <label
          class="form-label col-auto col-form-label"
        >
          Items per page:
        </label>
        <div
          class="col-auto"
        >
          <select
            class="form-select"
          >
            <option
              selected=""
              value="5"
            >
              5
            </option>
            <option
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="50"
            >
              50
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</div>
`;
